### PR TITLE
Add error type for combine keys + test and doc

### PR DESCRIPTION
**Pull request by Tibo-lg**
Trying to address the issues raised [here](https://github.com/rust-bitcoin/rust-secp256k1/pull/291#issuecomment-858397262):
* Add error type for combine_keys
* Add test with empty slice for combine_keys
* Document empty slice error condition for combine_keys

Deleted the `TweakCheckFailed` at the same time since it's not used but let me know if you'd prefer to have that in a separate commit/PR.

@real-or-random I was not sure if you wanted to change something in the fuzz_dummy module and I'm not so familiar with what it does so let me know if it does need a change.

### CODE REVIEW

**Comment by apoelstra** [comment]
This is a breaking change - let me check if we can push out a minor release before this.

**Comment by apoelstra** [comment]
Yes. Will need to wait for #305

**Comment by Tibo-lg** [comment]
> This is a breaking change

Is it because of deleting `TweakCheckFailed`? I don't think there is any hurry to merge this anyway but I can also just restore it.

**Comment by apoelstra** [comment]
Yes, though also because we added the other variant that replaces it.

File: src/key.rs [hunk 1]
```diff
@@ -395,11 +395,15 @@ impl PublicKey {
 
     /// Adds the keys in the provided slice together, returning the sum. Returns
     /// an error if the result would be the point at infinity, i.e. we are adding
-    /// a point to its own negation
+    /// a point to its own negation, or if the provided slice has no element in it.
     pub fn combine_keys(keys: &[&PublicKey]) -> Result<PublicKey, Error> {
```
Comment (dr-orlovsky) [inline src/key.rs:399]: I am not sure whether it is a good practice to return general error type when in fact only a single error subtype may be caused by the function. I think replacing `Result<_, Error>` on `Result<_, InvalidPublicKeySum>` will do a better job (the caller, if needed, may convert the error into general `Error` crate type).

Comment (Tibo-lg) [reply src/key.rs:399]: I can't speak for what is good practice or not, but looking at the rest of the code I feel like doing what you propose would be inconsistent with it, so if that's the way we want to go it might be better to make the change everywhere at once.

Comment (dr-orlovsky) [reply src/key.rs:399]: Yes, I also have seen that all functions in the lib used to return the single error type, which is a thing is not a good API, but this will certainly require distinct PR

**Review by dr-orlovsky — APPROVED** [review]

File: src/key.rs (hunk 1 above)
Comment (apoelstra) [reply src/key.rs:399]: If all of our functions returned distinct error types it would be an extremely unergonomic API, requring downstream users to wrap all of our individual types and implementing `From` on them to convert to their own error types.

**Review by apoelstra — DISMISSED** [review]
ack e2e1abf1a3f50daa049faa06dd7250b575f7a055

**Review by sgeisler — DISMISSED** [review]
utACK e2e1abf1a3f50daa049faa06dd7250b575f7a055

File: src/key.rs [hunk 2]
```diff
@@ -395,11 +395,15 @@ impl PublicKey {
...
         use core::mem::transmute;
         use core::i32::MAX;
 
+        if keys.len() < 1 {
+            return Err(InvalidPublicKeySum);
+        }
+
```
Comment (elichai) [inline src/key.rs:406]: Maybe it should be `!keys.is_empty() && keys.len() < i32::MAX as usize`? and then we can also remove the `debug_assert`

**Comment by elichai** [comment]
utACK 674cc79d8738fc079114311ebbdecaf992efcc39

**Review by apoelstra — APPROVED** [review]
ACK 674cc79d8738fc079114311ebbdecaf992efcc39
